subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/FileNameDerivingClassNameConverter.java

```diff
@@ -56,9 +56,20 @@ public FileNameDerivingClassNameConverter(SourceFileClassNameConverter delegate,
             return sourcePaths;
         }
 
+        Set<String> paths = fileExtensions.stream()
+            .map(fileExtension -> className.replace('.', '/') + fileExtension)
+            .collect(Collectors.toSet());
+
+        // Classes with $ may be inner classes
         int innerClassIdx = className.indexOf("$");
-        String baseName = innerClassIdx > 0 ? className.substring(0, innerClassIdx) : className;
-        return fileExtensions.stream().map(fileExtension -> baseName.replace('.', '/') + fileExtension).collect(Collectors.toSet());
+        if (innerClassIdx > 0) {
+            String baseName = className.substring(0, innerClassIdx);
+            paths.addAll(fileExtensions.stream()
+                .map(fileExtension -> baseName.replace('.', '/') + fileExtension)
```

Comment:
Would it be worth extracting a method which maps the FQCN to the relative path?

Reply:
I extracted `baseName.replace('.', '/') + fileExtension)` to a separate method.